File: 3Sum_4Sum/4sum.rs
fn backtrack(nums: &Vec<i32>, res: &mut Vec<Vec<i32>>, vec: &mut Vec<i32>, target: i32, value: i32, start: i32) {
    if vec.len() == 4 && value == target {
        res.push(vec.clone());
        return;
    }

    for i in start..nums.len() as i32 {
        if i > start && nums[i as usize] == nums[(i-1) as usize] {
            continue;
        }
        vec.push(nums[i as usize]);
        backtrack(nums, res, vec, target, nums[i as usize]+value, i+1);
        vec.pop();
    }
}

fn three_sum2(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    let mut vec = Vec::new();
    let mut nums_ = nums;
    nums_.sort();
    backtrack(&nums_, &mut res, &mut vec, target, 0, 0);
    res
}

fn three_sum1(nums: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    let mut res = Vec::new();
    let mut nums_ = nums;
    if nums_.len() < 4 {
        return res;
    }
    nums_.sort();

    for i in 0..nums_.len() - 3 {
        if i > 0 && nums_[i] == nums_[i-1] {
            continue;
        }
        for j in i+1..nums_.len() - 2 {
            if j > i+1 && nums_[j] == nums_[j-1] {
                continue;
            }
            let mut third = j + 1;
            let mut four = nums_.len() - 1;
            while third < four {
                let sum = nums_[i] + nums_[j] + nums_[third] + nums_[four];
                if sum < target {
                    third += 1;
                } else if sum > target {
                    four -= 1;
                } else {
                    res.push(vec![nums_[i], nums_[j], nums_[third], nums_[four]]);
                    while third < four && nums_[third] == nums_[third+1] {
                        third+= 1;
                    }
                    while third < four && nums_[four] == nums_[four-1] {
                        four -= 1;
                    }
                }
            }
        }
    }
    res
}

fn print_result(vecs: Vec<Vec<i32>>) {
    for vec in vecs {
        print!("[");
        for num in vec {
            print!("{} ", num);
        }
        println!("]");
    }
}

fn main() {
    let nums = vec![1, 0, -1, 0, -2, 2];
    let target = 0;

    let res1 = three_sum1(nums.clone(), target);
    print_result(res1);

    let res2 = three_sum2(nums.clone(), target);
    print_result(res2);
}

File: 3Sum_4Sum/4sum.py
def backtrack(nums, res, vec, target, value, start):
    if len(vec) == 4 and value == target:
        res.append(vec[:])
        return

    for i in range(start, len(nums)):
        if i > start and nums[i] == nums[i - 1]:
            continue
        vec.append(nums[i])
        backtrack(nums, res, vec, target, value + nums[i], i + 1)
        vec.pop()

def four_sum2(nums, target):
    res = []
    nums.sort()
    backtrack(nums, res, [], target, 0, 0)
    return res

def four_sum1(nums, target):
    res = []
    if len(nums) < 4:
        return res

    nums.sort()

    for i in range(len(nums) - 3):
        if i > 0 and nums[i] == nums[i - 1]:
            continue
        for j in range(i + 1, len(nums) - 2):
            if j > i + 1 and nums[j] == nums[j - 1]:
                continue
            third, four = j + 1, len(nums) - 1
            while third < four:
                sum_val = nums[i] + nums[j] + nums[third] + nums[four]
                if sum_val < target:
                    third += 1
                elif sum_val > target:
                    four -= 1
                else:
                    res.append([nums[i], nums[j], nums[third], nums[four]])
                    while third < four and nums[third] == nums[third + 1]:
                        third += 1
                    while third < four and nums[four] == nums[four - 1]:
                        four -= 1
    return res

def print_result(vecs):
    for vec in vecs:
        print(f"[{' '.join(map(str, vec))}]")

def main():
    nums = [1, 0, -1, 0, -2, 2]
    target = 0

    res1 = four_sum1(nums, target)
    #print_result(res1)

    #res2 = four_sum2(nums, target)
    #print_result(res2)

if __name__ == "__main__":
    main()
